File: 2023/cpp/day1/day1.cpp
#include <iostream>
#include <fstream>

std::string digits[20] =  {"1", "2", "3", "4", "5", "6", "7", "8", "",
                           "one", "two", "three", "four", "five",
                           "six", "seven", "eight", "nine", "ten"};

std::string findDigitsInString(std::string s){
    for(std::string digit : digits){
//      std::cout << digit << std::endl;
        size_t pos = s.find(digit);
        if (pos != std::string::npos){
            std::cout << digit << " at " << pos << std::endl;
        }
    } 
    return "";
}

std::string getCalibrationFromString(std::string s) {
    std::string left = "";
    std::string right = "";
    for(char& c : s) {
        if(isdigit(c)){
            left += c;
            break;
        }
    }
    for(int i = s.length(); i >= 0; i--){
        if(isdigit(s[i])){
            right += s[i];
            break;
        }
    }
    std::string rs = "";
    rs += left;
    rs += right;
    return (left + right);
}

void readDataFromFile(){
    int total = 0;
    std::string s; 
    std::ifstream f("input.txt");
    while(getline(f, s)){
        findDigitsInString(s);
//      int val = stoi(getCalibrationFromString(s));
//      total += val;
//      std::cout << val << std::endl;
    }
    std::cout << total << std::endl;
    f.close();
    return;
}

int main() {
    readDataFromFile();
    return 0;
}

File: 2023/py/day1/day1.py
digits = [
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'one',
    'two',
    'three',
    'four',
    'five',
    'six',
    'seven',
    'eight',
    'nine',
]

digit_maps = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9'
}

def find_digitloc_in_line(line):
    first = [999, '']
    last = [-1, '']
    for d in digits:
        lloc = line.find(d)
        rloc = line.rfind(d)
        if lloc != -1 and lloc < first[0]:
            first = lloc, d
        if rloc > last[0]:
            last = rloc, d
#    print(first, last)
    if not first[1].isdigit():
        first = digit_maps[first[1]]
    else:
        first = first[1]
    if not last[1].isdigit():
        last = digit_maps[last[1]]
    else:
        last = last[1]
#    print(first + last)
    return first + last

def read_input_from_file(file):
    with open(file) as f:
        input = f.read().split('\n')
        print(input)
        return(input)

def get_calibration_value_from_line(line):
    sum = ''
    for char in line:
        if char.isdigit():
            sum += char
            break
    for char in reversed(line):
        if char.isdigit():
            sum += char
            break
    print(sum)
    return(sum)

if __name__ == '__main__':
    input = read_input_from_file('input.txt')
#    print(find_digitloc_in_line('sdpgz3five4seven6fiveh'))
    sum = 0
    for i, line in enumerate(input):
        val = int(find_digitloc_in_line(line))
        print(i+1, val)
        sum += val
    print(sum)
